smart-contracts/contracts/Relayer.sol

```diff
@@ -2,87 +2,268 @@
 pragma solidity ^0.8.0;
 
 import "./NilTokenManager.sol";
+import "./IterableMapping.sol";
+import "../system/console.sol";
+
+// Dirty hack to run the tests with lesser number of shards
+uint32 constant _N_SHARDS = 6;
 
 /**
  * @title Relayer
  * @dev This contract facilitates relaying transactions, handling responses, and managing token credits.
+ * It contains a message queue for asynchronous cross-shard communication.
  */
 contract Relayer {
-
-    struct Transaction {
+    using IterableMapping for IterableMapping.Map;
+
+    // Structure for cross-shard messages
+    struct Message {
+        uint64 id;                   // message ID
+        uint64 seqno;                // Sequence number
```

Comment:
https://docs.soliditylang.org/en/latest/internals/layout_in_storage.html because layout of storage it's better to forwrarkind and requestId with bool be on the up and they will take one word instead of 2 words. It could be a little help to reduce a size of struct.